### PR TITLE
i18n support for file collections - closes #4483

### LINKED ISSUE #4483: i18n: Add support for file collections
We are currently on a forked version of the [first PR of the internationalization track](https://github.com/netlify/netlify-cms/pull/2988) where [we added support for list collections](https://github.com/reimertz/netlify-cms/commit/8ebe03a08e4aaab8fa038caf226dd425c8e7b1b2).

But since it was decided to rework above implementation and use another architecture, our solution became obsolete and are now stuck on that fork. 😅

So naturally, we've started looking into contributing by adding support for file collections but would like to ask @erezrokah if there was any other reason for not adding this feature except from scoping the PR?  

Something else we should thing about when we add the feature? 
Any pointers / ideas / etc before we go deep here would be very helpful.

### PR DESCRIPTION
**Pull request by gewfy**
**Summary**
This adds i18n support for file collections as per #4483

**Test plan**
Fixed unit tests:
- [x] Removed `config > applyDefaults > i18n > should throw when i18n is set on files collection`
- [x] Added `config > applyDefaults > i18n > should set i18n value to translate on field when i18n=true for field in files collection`

Working in development with `i18n: true` on the `Settings > Site Settings` file collection. 
image

**A picture of a cute animal (not mandatory but encouraged)**
Worlds smallest horse!
image

### CODE REVIEW

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks @gewfy.
This enables the translation UI for file collections but doesn't handle persisting the content.

Can you please share an example configuration and the resulting content files of saving a file with `i18n` support? How would those files be named and structured in the repo?

**Comment by gewfy** [comment]
@erezrokah Ok so if I understand you correctly. Basically having a file collection configured without the actual file being there? I see that this errors now. Will see if I can fix that.

We're thinking about adding `i18n Posts` and `i18n Settings` collections to the `dev-test.html`. Is that a good idea you think?

**Comment by gewfy** [comment]
@erezrokah I've limited this change to only support `single_file` structures since (I'm guessing) that adding support for `multiple_folders` and `multiple_files` would require a different way of configuring files collections.

What I've done:
* Made sure that `single_file` files collections is editable and persists for both existing and non-existing files
* Throw if the top level or collection level i18n structure is set to anything else than `single_file`
* Tweaked the [i18n limitations](https://www.netlifycms.org/docs/beta-features/#limitations) parts of the docs
* Added `i18n Posts` and `i18n Settings` to `dev-test.html` for easier testing

Good enough trade-off?

File: dev-test/config.yml [hunk 1]
```diff
@@ -6,6 +6,11 @@ site_url: https://example.com
 publish_mode: editorial_workflow
 media_folder: assets/uploads
 
+i18n:
```
Comment (erezrokah) [inline dev-test/config.yml:9]: Can we revert these changes and the ones to `index.html`?
This is what gets deployed to 
https://cms-demo.netlify.com/#/collections/posts
See https://deploy-preview-4487--cms-demo.netlify.app/#/collections/i18n_settings

I don't think we want it as a part of the demo site (yet)

File: packages/netlify-cms-core/src/actions/__tests__/config.spec.js [hunk 2]
```diff
@@ -634,7 +634,7 @@ describe('config', () => {
         ).toEqual({ structure: 'multiple_folders', locales: ['en', 'fr'], default_locale: 'fr' });
       });
 
-      it('should throw when i18n is set on files collection', () => {
+      it('should throw when i18n structure is not single_file on files collection', () => {
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/__tests__/config.spec.js:637]: Can we explicitly test for an error being thrown when structure is one of `multiple_folders, multiple_files`?

File: packages/netlify-cms-core/src/actions/config.js [hunk 3]
```diff
@@ -90,14 +90,33 @@ const setI18nDefaults = (i18n, collection) => {
...
-      collection = collection.set('fields', traverseFields(collection.get('fields'), setI18nField));
+      if (collection.has('fields')) {
+        collection = collection.set(
+          'fields',
+          traverseFields(collection.get('fields'), setI18nField),
+        );
+      }
+
+      if (collection.has('files')) {
+        collection = collection.set('files', traverseFields(collection.get('files'), setI18nField));
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.js:101]: `traverseFields` fields expects a list of `fields` and not of files.

File: packages/netlify-cms-core/src/actions/config.js [hunk 4]
```diff
@@ -90,14 +90,33 @@ const setI18nDefaults = (i18n, collection) => {
...
+      collection = collection.set(
+        'fields',
+        traverseFields(collection.get('fields'), field => field.delete(I18N)),
+      );
+    }
+
+    if (collection.has('files')) {
+      collection = collection.set(
+        'files',
+        traverseFields(collection.get('files'), field => field.delete(I18N)),
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.js:117]: `traverseFields` fields expects a list of `fields` and not of files.

File: packages/netlify-cms-core/src/actions/config.js [hunk 5]
```diff
@@ -231,9 +250,17 @@ export function applyDefaults(config) {
 
           const files = collection.get('files');
           if (files) {
-            if (i18n && collection.has(I18N)) {
-              throw new Error('i18n configuration is not supported for files collection');
+            if (
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.js:253]: Can we extract this logic to a function, e.g. `isI18nAllowed`?
Also I think setting:
```yaml
file:
  i18n:
    structure: multiple_folders
```
Would pass this check

File: packages/netlify-cms-core/src/actions/config.js [hunk 6]
```diff
@@ -245,9 +272,11 @@ export function applyDefaults(config) {
                   'fields',
                   traverseFields(file.get('fields'), setDefaultPublicFolder),
                 );
+                file = setI18nDefaults(i18n, file);
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/actions/config.js:275]: files should inherit collection level configuration. I think calling `setI18nDefaults` this way will inherit top level configuration.

This will require re-using some of the logic from here:
https://github.com/netlify/netlify-cms/blob/519cb2d4c2db729d2643c9116f93656b6a9dba23/packages/netlify-cms-core/src/actions/config.js#L74

**Review by erezrokah — CHANGES_REQUESTED** [review]
Thanks @gewfy, enabling `single_file` for file collections makes sense.
Added a few comments.
I think the changes will require splitting https://github.com/netlify/netlify-cms/blob/519cb2d4c2db729d2643c9116f93656b6a9dba23/packages/netlify-cms-core/src/actions/config.js#L71 

1. Set the defaults for the collection based on the top level config.
2. Set the defaults for each file based on the collection config.
3. Set the defaults for each field based on the collection/file.

**Comment by erezrokah** [comment]
Replaced by https://github.com/netlify/netlify-cms/pull/4634 as I couldn't push to the branch of this PR
